core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -421,7 +423,32 @@ class KafkaApis(val requestChannel: RequestChannel,
 
     // the callback for sending a fetch response
     def sendResponseCallback(responsePartitionData: Map[TopicAndPartition, FetchResponsePartitionData]) {
-      val mergedResponseStatus = responsePartitionData ++ unauthorizedResponseStatus
+
+      val convertedResponseStatus =
+        // Need to down-convert message when consumer only takes magic value 0.
+        if (fetchRequest.versionId <= 1) {
+          responsePartitionData.map({ case (tp, data) =>
+            tp -> {
+              // We only do down-conversion when:
+              // 1. The message format version configured for the topic is using magic value > 0, and
+              // 2. The message set contains message whose magic > 0
+              // This is to reduce the message format conversion as much as possible. The conversion will only occur
+              // when new message format is used for the topic and we see an old request.
+              // Please notice that if the message format is changed from a higher version back to lower version this
+              // test might break because some messages in new message format can be delivered to consumer without
+              // format down conversion.
+              if (replicaManager.getMessageFormatVersion(tp).exists(_ > 0) &&
```

Comment:
Hi Jun, I ran the following experiment with the configuration check removed. i.e. always verify messages for fetch request v1.

1. Start a new broker with message.format.version=v0.
2. produce 3000000 messages to topic with 128 partition using console producer.
3. Consume the messages using console consumer in this RB (sending v2 fetch request, so no message format verification needed.)
4. Consume the messages using console consumer in current trunk (sending v1 fetch request, so message verification is needed)

The log below prints out the time cost on the verification code block:
```
[2016-02-12 10:33:33,976] INFO [KafkaApi-0] Avg time taken = 1163 (kafka.server.KafkaApis)
[2016-02-12 10:33:51,126] INFO [KafkaApi-0] Avg time taken = 933 (kafka.server.KafkaApis)
[2016-02-12 10:34:07,205] INFO [KafkaApi-0] Avg time taken = 907 (kafka.server.KafkaApis)
[2016-02-12 10:34:23,627] INFO [KafkaApi-0] Avg time taken = 962 (kafka.server.KafkaApis)
[2016-02-12 10:35:48,784] INFO [KafkaApi-0] Avg time taken = 7914977 (kafka.server.KafkaApis)
[2016-02-12 10:36:05,932] INFO [KafkaApi-0] Avg time taken = 18795324 (kafka.server.KafkaApis)
[2016-02-12 10:36:23,004] INFO [KafkaApi-0] Avg time taken = 20152849 (kafka.server.KafkaApis)
[2016-02-12 10:36:39,973] INFO [KafkaApi-0] Avg time taken = 19948298 (kafka.server.KafkaApis)
[2016-02-12 10:36:57,072] INFO [KafkaApi-0] Avg time taken = 19525645 (kafka.server.KafkaApis)
```
The first 4 lines are from v2 requests. The fifth line is from v1 and v2 mixed. The last 4 lines are from v1 requests. Time unit is nano seconds.

It seems the traversal cost is expensive.